src/main.cc

```diff
@@ -1951,55 +2046,25 @@ watch_child(char *argv[])
             waitFlag = WNOHANG;
         PidStatus status;
         pid = WaitForAnyPid(status, waitFlag);
+        getCurrentTime();
 
         // check for a stopped kid
-        Kid* kid = pid > 0 ? TheKids.find(pid) : NULL;
-        if (kid) {
+        if (Kid *kid = pid > 0 ? TheKids.find(pid) : nullptr)
             kid->stop(status);
-            if (kid->calledExit()) {
-                syslog(LOG_NOTICE,
-                       "Squid Parent: %s process %d exited with status %d",
-                       kid->name().termedBuf(),
-                       kid->getPid(), kid->exitStatus());
-            } else if (kid->signaled()) {
-                syslog(LOG_NOTICE,
-                       "Squid Parent: %s process %d exited due to signal %d with status %d",
-                       kid->name().termedBuf(),
-                       kid->getPid(), kid->termSignal(), kid->exitStatus());
-            } else {
-                syslog(LOG_NOTICE, "Squid Parent: %s process %d exited",
-                       kid->name().termedBuf(), kid->getPid());
-            }
-            if (kid->hopeless()) {
-                syslog(LOG_NOTICE, "Squid Parent: %s process %d will not"
-                       " be restarted due to repeated, frequent failures",
-                       kid->name().termedBuf(), kid->getPid());
-            }
-        } else if (pid > 0) {
+        else if (pid > 0)
             syslog(LOG_NOTICE, "Squid Parent: unknown child process %d exited", pid);
-        }
+
+        masterCheckAndBroadcastSignals();
+        masterMaintainKidRevivalSchedule();
 
         if (!TheKids.someRunning() && !TheKids.shouldRestartSome()) {
             leave_suid();
             // XXX: Master process has no main loop and, hence, should not call
             // RegisteredRunner::startShutdown which promises a loop iteration.
             RunRegisteredHere(RegisteredRunner::finishShutdown);
             enter_suid();
-
-            if (TheKids.someSignaled(SIGINT) || TheKids.someSignaled(SIGTERM)) {
-                syslog(LOG_ALERT, "Exiting due to unexpected forced shutdown");
-                exit(EXIT_FAILURE);
-            }
-
-            if (TheKids.allHopeless()) {
-                syslog(LOG_ALERT, "Exiting due to repeated, frequent failures");
-                exit(EXIT_FAILURE);
-            }
-
-            exit(EXIT_SUCCESS);
+            MasterExit();
```

Comment:
Please `s/MasterExit/masterExit/` for consistency sake. We should eventually capitalize all of the static `master*()` functions or, better, move them into a new class, but either is outside this project scope.